Signing powers

vzotova opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
Separates operator key and signing key

**Issues fixed/closed:**
> - Fixes #...

Depends on https://github.com/nucypher/nucypher-contracts/pull/436

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
Open question (@KPrasch @derekpierre ): is ThresholdSIgningPower default power for Ursula, and do we want to have default constructor (no parameters, random generation) for it?

KPrasch committed:
initial dedicated signing wallet derivation and hookup

vzotova committed:
Follow up work with tests

vzotova committed:
Extract signer address from operator

derekpierre commented on nucypher/crypto/powers.py:
```diff
@@ -236,6 +237,33 @@ def sign_transaction(self, transaction_dict: dict) -> bytes:
         return self._signer.sign_transaction(transaction_dict=transaction_dict)
 
 
+class ThresholdSigningPower(TransactingPower):
```
@vzotova , @KPrasch can this just be a subclass of `SigningPower` instead of `TransactingPower` and added as a default crypto power up for Ursula.

KPrasch replied:
Due to the legacy of powers in which "signing" was imagined to be a strictly P2P notion, SigningPower is built around a different cryptographic paradigm.  `TransactingPower` is already nearly purpose-built for the purpose of this PR (an ethereum key wrapper), although I can understand the naming confusion.  In terms of default powers, since TransactingPower (and TLSHostingPower) require runtime-context, we currently do not have a design that supports them being defaults.  I'd regard this as a design imperfection, that being said, TransactingPower has enjoyed special treatment that can be repeated for the `ThresholdSigningPower`.

TLDR; IMO it does not make sense to make `ThresholdSigningPower` a subclass of `SigningPower` and in order to avoid design upheveal `ThresholdSigningPower` can be implemented as a special case, not a default power.

vzotova committed:
Adds newsfragment

derekpierre committed:
Move instantiation of trackers to the end of the Operator.__init__ method since these take operators as parameters; ensure that the operator is fully initialized before passing along.

derekpierre committed:
Use InMemorySigner with ThresholdSigningPower since it is only derived from the Keystore and not from any keyfile or external user input.
Operator will instantiate the default power up for ThresholdSigningPower instead of having it passed in as a constructor argument.

derekpierre committed:
Update test fixture to only pass the InMemorySigner when instantiating the test ThresholdSigningPower.

derekpierre commented:
@vzotova, just pushed the commits we chatted about. The earlier commits in the PR are unsigned, and they will need to be signed for the PR to be merged.

Also, does it make sense to look into merging https://github.com/nucypher/nucypher-contracts/pull/436 before this PR since it depends on those contract changes? Not an issue, but if the contracts PR gets merged, then you don't need the updates to this `ape-config.yml` file in this PR to update the contracts repos/branch to use. Of course, if you prefer to merge this PR first, we can always make that update at a later time.

vzotova commented:
> @vzotova, just pushed the commits we chatted about. The earlier commits in the PR are unsigned, and they will need to be signed for the PR to be merged.
> 
> Also, does it make sense to look into merging [nucypher/nucypher-contracts#436](https://github.com/nucypher/nucypher-contracts/pull/436) before this PR since it depends on those contract changes? Not an issue, but if the contracts PR gets merged, then you don't need the updates to this `ape-config.yml` file in this PR to update the contracts repos/branch to use. Of course, if you prefer to merge this PR first, we can always make that update at a later time.

good point, I'd merge PR in nucypher-contracts first

cygnusv commented on nucypher/network/concurrency.py:
```diff
@@ -205,8 +205,8 @@ def worker(
-        # successes is of type Dict[ChecksumAddress, Tuple[ChecksumAddress, ThresholdSignatureResponse]]
+        # sort successes by signer address
+        # successes is of type Dict[ChecksumAddress, Tuple[ChecksumAddress, SignatureResponse]]
         successes = OrderedDict(sorted(successes.items(), key=lambda item: item[1][0]))
```
Just to point out that this sorts by signer address _as strings_. It doesn't seem that this has any effect, but in case it matters for something else.

derekpierre replied:
Good catch - @vzotova, that was my prior mistake separate from this PR. Can you update the code to mimic the ordering of addresses here - https://github.com/nucypher/nucypher/blob/v7.7.x/tests/acceptance/actors/test_dkg_ritual.py#L173

vzotova replied:
✔️

cygnusv commented on nucypher/crypto/keystore.py:
```diff
@@ -46,6 +48,7 @@
 _RITUALISTIC_INFO = __INFO_BASE + b"ritualistic"
 _THRESHOLD_REQUEST_DECRYPTING_INFO = __INFO_BASE + b"threshold_request_decrypting"
 _TLS_INFO = __INFO_BASE + b"tls"
+_THRESHOLD_SIGNING_INFO = __INFO_BASE + b"threshold_signing"
```
Maybe let's suffix this with ECDSA: `b"threshold_signing_ecdsa"`. See my other comments in `powers.py`

vzotova replied:
✔️

cygnusv commented on nucypher/crypto/powers.py:
```diff
@@ -236,6 +241,21 @@ def sign_transaction(self, transaction_dict: dict) -> bytes:
+class ThresholdSigningPower(TransactingPower):
+    """A power that is dedicated to sign threshold cryptography requests."""
+
+    KEY_SIZE = 32  # must be 32 bytes long
```
Noting that this is ECDSA-specific.

vzotova replied:
✔️

cygnusv commented on tests/acceptance/ape-config.yaml:
```diff
@@ -5,8 +5,8 @@ plugins:
-    github: nucypher/nucypher-contracts
-    ref: signing
+    github: vzotova/nucypher-contracts
+    ref: operator-signer
```
We should change this, right?

derekpierre replied:
This will be changed back once https://github.com/nucypher/nucypher-contracts/pull/436 is merged.

derekpierre replied:
^ Reminder about this once the `nucypher-contracts` PR is merged. It has sufficient approvals already, so feel free to merge.

cygnusv commented on tests/acceptance/agents/test_signing_coordinator_agent.py:
```diff
@@ -201,6 +211,9 @@ def test_post_signature(
-    assert deployed_multisig.getSigners() == cohort_operators
+    assert (
+        deployed_multisig.getSigners()
+        == accounts.unassigned_accounts[: len(cohort_operators)]
```
Perhaps `accounts.unassigned_accounts[: len(cohort_operators)]` should be made into a fixture. Same for L49.

vzotova replied:
✔️

cygnusv commented on nucypher/blockchain/eth/models.py:
```diff
@@ -224,12 +224,14 @@ def make_signers(
                 operator=ChecksumAddress(data[1]),
-                signature=bytes(data[2]),
+                signerAddress=ChecksumAddress(data[2]),
+                signature=bytes(data[3]),
```
Does it make sense to recover the address from the signature? Since this is ECDSA anyway

vzotova replied:
it mimics what happens in contracts, so after finishing discussion about what we store and what relationship between signer<->operator - we can return to this one. If everything stays as now - I'd remove signature from the struct in contract

vzotova replied:
✔️

vzotova committed:
Apply RFC from #3659.

Co-authored-by: David Núñez <david@nucypher.com>

vzotova committed:
Adapt latest changes in contract

cygnusv commented on newsfragments/3659.dev.rst:
```diff
@@ -0,0 +1 @@
+Dedicated signing key
```
```suggestion
Dedicated signing key for Action Control multisig messages.
```

derekpierre replied:
^ Did this commit get lost?

vzotova replied:
thanks!
returned changes back

cygnusv commented on nucypher/blockchain/eth/actors.py:
```diff
@@ -1275,7 +1279,7 @@ def perform_post_signature(
         if async_tx:
             self.log.info(
-                f"Active signing ritual in progress: {self.transacting_power.account} has submitted tx "
+                f"Active signing ritual in progress: {self.threshold_signing_power.account} has submitted tx "
```
The logging intention here is ambiguous: Do we want to log that an operator account submitted a TX or that a signer produced a signature? As far as I understand, the signer doesn't submit TXs, but operators do.

derekpierre replied:
^ one more spot here @vzotova .

vzotova replied:
✔️

derekpierre commented on nucypher/blockchain/eth/actors.py:
```diff
@@ -1211,7 +1215,7 @@ def resubmit_tx():
 
             self.log.info(
-                f"{self.transacting_power.account[:8]} resubmitted a new async tx {async_tx.id} "
+                f"{self.threshold_signing_power.account[:8]} resubmitted a new async tx {async_tx.id} "
```
Careful - this is still technically the operator (`transacting_power`) submitting the tx. Should we change this back to `transacting_power`?

vzotova replied:
✔️

derekpierre commented on nucypher/network/concurrency.py:
```diff
@@ -185,8 +185,8 @@ def worker(
-                    operator_address = node_or_sprout.operator_address
-                    return operator_address, response
+                    signer_address = response.signer
+                    return signer_address, response
```
Signer address is already in the response (`SignatureResponse` object), so let's just return the response instead of the tuple.

vzotova replied:
✔️

vzotova committed:
Apply RFC from #3659.

Co-authored-by: David Núñez <david@nucypher.com>
Co-authored-by: derekpierre <derek.pierre@gmail.com>

derekpierre reviewed (approved):
🎸 - nice work!

Looks like the newsfragment commit may have gotten lost in the shuffle - all good otherwise,

vzotova committed:
Changing back ape-config.yaml
